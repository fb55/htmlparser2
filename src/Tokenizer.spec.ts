import { Tokenizer } from ".";

function tokenize(str: string) {
    const log: unknown[][] = [];
    const tokenizer = new Tokenizer(
        {},
        new Proxy({} as any, {
            get(_, prop) {
                return (...args: unknown[]) => log.push([prop, ...args]);
            },
        })
    );

    tokenizer.write(str);
    tokenizer.end();

    return log;
}

describe("Tokenizer", () => {
    describe("should support self-closing special tags", () => {
        it("for self-closing script tag", () => {
            expect(tokenize("<script /><div></div>")).toMatchSnapshot();
        });
        it("for self-closing style tag", () => {
            expect(tokenize("<style /><div></div>")).toMatchSnapshot();
        });
        it("for self-closing title tag", () => {
            expect(tokenize("<title /><div></div>")).toMatchSnapshot();
        });
    });

    describe("should support standard special tags", () => {
        it("for normal script tag", () => {
            expect(tokenize("<script></script><div></div>")).toMatchSnapshot();
        });
        it("for normal style tag", () => {
            expect(tokenize("<style></style><div></div>")).toMatchSnapshot();
        });
        it("for normal sitle tag", () => {
            expect(tokenize("<title></title><div></div>")).toMatchSnapshot();
        });
    });

    it("should not lose data when pausing", () => {
        const log: unknown[][] = [];
        const tokenizer = new Tokenizer(
            {},
            new Proxy({} as any, {
                get(_, prop) {
                    return (...args: unknown[]) => {
                        if (prop === "ontext") {
                            tokenizer.pause();
                        }
                        log.push([prop, ...args]);
                    };
                },
            })
        );

        tokenizer.write("&amp; it up!");
        tokenizer.resume();
        tokenizer.resume();
        tokenizer.end();

        expect(log).toMatchSnapshot();
    });
});
