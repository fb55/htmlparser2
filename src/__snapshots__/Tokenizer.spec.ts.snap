// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenizer should not lose data when pausing 1`] = `
[
  [
    "ontextentity",
    38,
  ],
  [
    "ontext",
    5,
    12,
  ],
  [
    "onend",
  ],
]
`;

exports[`Tokenizer should support self-closing special tags for self-closing script tag 1`] = `
[
  [
    "onopentagname",
    1,
    7,
  ],
  [
    "onselfclosingtag",
    9,
  ],
  [
    "onopentagname",
    11,
    14,
  ],
  [
    "onopentagend",
    14,
  ],
  [
    "onclosetag",
    17,
    20,
  ],
  [
    "onend",
  ],
]
`;

exports[`Tokenizer should support self-closing special tags for self-closing style tag 1`] = `
[
  [
    "onopentagname",
    1,
    6,
  ],
  [
    "onselfclosingtag",
    8,
  ],
  [
    "onopentagname",
    10,
    13,
  ],
  [
    "onopentagend",
    13,
  ],
  [
    "onclosetag",
    16,
    19,
  ],
  [
    "onend",
  ],
]
`;

exports[`Tokenizer should support self-closing special tags for self-closing title tag 1`] = `
[
  [
    "onopentagname",
    1,
    6,
  ],
  [
    "onselfclosingtag",
    8,
  ],
  [
    "onopentagname",
    10,
    13,
  ],
  [
    "onopentagend",
    13,
  ],
  [
    "onclosetag",
    16,
    19,
  ],
  [
    "onend",
  ],
]
`;

exports[`Tokenizer should support standard special tags for normal script tag 1`] = `
[
  [
    "onopentagname",
    1,
    7,
  ],
  [
    "onopentagend",
    7,
  ],
  [
    "onclosetag",
    10,
    16,
  ],
  [
    "onopentagname",
    18,
    21,
  ],
  [
    "onopentagend",
    21,
  ],
  [
    "onclosetag",
    24,
    27,
  ],
  [
    "onend",
  ],
]
`;

exports[`Tokenizer should support standard special tags for normal sitle tag 1`] = `
[
  [
    "onopentagname",
    1,
    6,
  ],
  [
    "onopentagend",
    6,
  ],
  [
    "onclosetag",
    9,
    14,
  ],
  [
    "onopentagname",
    16,
    19,
  ],
  [
    "onopentagend",
    19,
  ],
  [
    "onclosetag",
    22,
    25,
  ],
  [
    "onend",
  ],
]
`;

exports[`Tokenizer should support standard special tags for normal style tag 1`] = `
[
  [
    "onopentagname",
    1,
    6,
  ],
  [
    "onopentagend",
    6,
  ],
  [
    "onclosetag",
    9,
    14,
  ],
  [
    "onopentagname",
    16,
    19,
  ],
  [
    "onopentagend",
    19,
  ],
  [
    "onclosetag",
    22,
    25,
  ],
  [
    "onend",
  ],
]
`;
